src/main/java/org/broadinstitute/hellbender/metrics/MultiLevelCollectorReducible.java

```diff
@@ -0,0 +1,428 @@
+            throw new IllegalArgumentException("MultiLevelCollectors must be the same size to be combined");
+        }
+
+        // Even though we'll visit the allReadsDistributor below, since it doesn't actually
+        // keep it's collector in the normal distributor map, we need to manually merge the
+        // allReadCollector here
+        if (allReadCollector != null) {
+            this.allReadCollector = combineUnit(this.allReadCollector, source.allReadCollector);
+        }
+
+        if (this.outputOrderedDistributors.size() != source.outputOrderedDistributors.size()) {
+            throw new IllegalArgumentException("MultiLevelCollectors must be the same size to be combined");
+        }
+
+        for (int i = 0; i < this.outputOrderedDistributors.size(); i++) {
+            this.outputOrderedDistributors.get(i).combine(
+                    source.outputOrderedDistributors.get(i), this::combineUnit);
+        }
+    }
+
+    /**
+     * Combined two UNIT_COLLECTOR objects into a single object representing the combined collectors.
+     *
+     * Subclasses must implement this method. Note that it is acceptable for implementations to be
+     * destructive and combine objects by mutating and returning one of the input objects.
+     *
+     * @param c1 first unit collector
+     * @param c2 second unit collector
+     * @return UNIT_COLLECTOR resulting from the combination of the two input collectors
+     */
+    public abstract UNIT_COLLECTOR combineUnit(UNIT_COLLECTOR c1, UNIT_COLLECTOR c2);
+
+    /**
+     * Distributors group PerUnitMetricCollectors based on a AccumulationLevel.  Their structure mimics
+     * PerUnitMetricCollectors but instead of adding records to metrics they identify which
+     * PerUnitMetricCollector should receive a specific record and distribute records on to the that collector
+     *
+     * There were will be 0 or 1 Distributors for each of the following MetriAcummulationLevels:
+     * ALL_READS, SAMPLE, LIBRARY, READ_GROUP
+     */
+    private abstract class Distributor implements Serializable {
+        private static final long serialVersionUID = 1L;
+
+        //A Map mapping the key for a specific record (as determined by getKey) to the appropriate collector
+        private final Map<String, UNIT_COLLECTOR> collectors;
+
+        //Given a SAMReadGroupRecord, return the key that identifies the collector for the corresponding SAMRecord
+        protected abstract String getKey(final SAMReadGroupRecord rg);
+
+        //Make a PerUnitMetricCollector for this given Distributor
+        protected abstract UNIT_COLLECTOR makeCollector(final SAMReadGroupRecord rg);
+
+        protected abstract UNIT_COLLECTOR makeUnknownCollector();
+
+        public Distributor(final List<SAMReadGroupRecord> rgRecs) {
+            collectors = new LinkedHashMap<>();
+            for(final SAMReadGroupRecord rg : rgRecs) {
+                final String key = getKey(rg);
```

Comment:
the getKey and makeCollector methods are called during object initialization, which may result in hard to detect bugs because methods typically assume that the object has been created and initialized